Source/DafnyDriver/DafnyDoc.cs

```diff
@@ -0,0 +1,1142 @@
+        Directory.CreateDirectory(outputdir);
+      }
+
+      // Check writable
+      try {
+        File.Create(outputdir + "/index.html").Dispose();
+      } catch (Exception) {
+        reporter.Error(MessageSource.Documentation, Token.NoToken, "Insufficient permission to create output files in " + outputdir);
+        return DafnyDriver.ExitValue.DAFNY_ERROR;
+      }
+      // Generate all the documentation
+      exitValue = new DafnyDoc(dafnyProgram, reporter, options, outputdir).GenerateDocs(dafnyFiles);
+    }
+    return exitValue;
+  }
+
+  public Program DafnyProgram;
+  public ErrorReporter Reporter;
+  public DafnyOptions Options;
+  public string Outputdir;
+  public Dictionary<string, string> nameIndex = new Dictionary<string, string>();
+
+  public DafnyDoc(Program dafnyProgram, ErrorReporter reporter, DafnyOptions options, string outputdir) {
+    this.DafnyProgram = dafnyProgram;
+    this.Reporter = reporter;
+    this.Options = options;
+    this.Outputdir = outputdir;
+  }
+
+  public DafnyDriver.ExitValue GenerateDocs(IList<DafnyFile> dafnyFiles) {
+    try {
+      var modDecls = new List<LiteralModuleDecl>();
+      var rootModule = DafnyProgram.DefaultModule;
+      var decls = rootModule.ModuleDef.TopLevelDecls.Select(d => !(d is LiteralModuleDecl));
+      CollectDecls(rootModule, modDecls);
+      WriteTOC(modDecls);
+      foreach (var m in modDecls) {
+        WriteModule(m, dafnyFiles);
+      }
+      WriteIndex();
+      return DafnyDriver.ExitValue.SUCCESS;
+    } catch (Exception e) {
+      // This is a fail-safe backstop so that dafny itself does not crash
+      Reporter.Error(MessageSource.Documentation, DafnyProgram.DefaultModule, $"Unexpected exception while generating documentation: {e.Message}\n{e.StackTrace}");
+      return DafnyDriver.ExitValue.DAFNY_ERROR;
+    }
+  }
+
+  /** Recursively computes a list of all the LiteralModuleDecls declared in the program */
+  public void CollectDecls(LiteralModuleDecl mod, List<LiteralModuleDecl> modDecls) {
+    modDecls.Add(mod);
+    foreach (var d in mod.ModuleDef.TopLevelDecls) {
+      if (d is LiteralModuleDecl litmod) {
+        CollectDecls(litmod, modDecls);
+      }
+    }
+  }
+
+  /** Writes a doc page for the given module */
+  public void WriteModule(LiteralModuleDecl module, IList<DafnyFile> dafnyFiles) {
+    var moduleDef = module.ModuleDef;
+    var fullName = moduleDef.FullDafnyName;
+    var fullNLName = fullName;
+    if (moduleDef.IsDefaultModule) {
+      nameIndex.Add(RootNLName + " " + nameIndex.Count + " " + RootName, "module " + Link(RootName, RootNLName));
+      fullName = RootName;
+      fullNLName = RootNLName;
+    } else {
+      AddToIndexF(module.Name, fullName, "module");
+    }
+    var defaultClass = moduleDef.TopLevelDecls.First(d => d is ClassDecl cd && cd.IsDefaultClass) as ClassDecl;
+
+    string filename = Outputdir + "/" + fullName + ".html";
+    using StreamWriter file = new(filename);
```

Comment:
Could you separate the TextWriter/StreamWriter creation, which includes determining the filename) and the actual writing code?